Algorithm/QCAlgorithm.Trading.cs

```diff
@@ -604,36 +607,32 @@ public int CalculateOrderQuantity(Symbol symbol, decimal target)
             // determine the unit price in terms of the account currency
             var unitPrice = new MarketOrder(symbol, 1, UtcTime).GetValue(security);
 
-            // define lower and upper thresholds for the iteration
-            var lowerThreshold = targetOrderValue - unitPrice / 2;
-            var upperThreshold = targetOrderValue + unitPrice / 2;
+            // calculate the total margin available
+            var marginRemaining = Portfolio.GetMarginRemaining(symbol, direction);
+            if (marginRemaining <= 0) return 0;
 
-            // continue iterating while  we're still not within the specified thresholds
-            var iterations = 0;
+            // continue iterating while we do not have enough margin for the order
+            decimal marginRequired;
             var orderQuantity = 0;
-            decimal orderValue = 0;
-            while ((orderValue < lowerThreshold || orderValue > upperThreshold) && iterations < 10)
-            {
-                // find delta from where we are to where we want to be
-                var delta = targetOrderValue - orderValue;
-                // use delta value to compute a change in quantity required
-                var deltaQuantity = (int)(delta / unitPrice);
 
-                orderQuantity += deltaQuantity;
+            do
+            {
+                if (orderQuantity == 0)
+                {
+                    // compute the initial order quantity
+                    orderQuantity = (int)(targetOrderValue / unitPrice);
+                }
+                else
+                {
+                    // decrease the order quantity
+                    orderQuantity--;
+                }
 
-                // recompute order fees
+                // calculate the margin required for the order
                 var order = new MarketOrder(security.Symbol, orderQuantity, UtcTime);
-                var fee = security.FeeModel.GetOrderFee(security, order);
-
-                orderValue = Math.Abs(order.GetValue(security)) + fee;
+                marginRequired = security.MarginModel.GetInitialMarginRequiredForOrder(security, order);
 
-                // we need to add the fee in as well, even though it's not value, it's still a cost for the transaction
-                // and we need to account for it to be sure we can make the trade produced by this method, imagine
-                // set holdings 100% with 1x leverage, but a high fee structure, it quickly becomes necessary to include
-                // otherwise the result of this function will be inactionable.
-
-                iterations++;
-            }
+            } while (marginRequired > marginRemaining || marginRequired > targetOrderValue);
```

Comment:
I'm curious as to how many times this iterates for common cases, since it looks like it's decrementing one at a time instead of estimating how many to decrement, so for large order fees (fees that are large compared to price of single share) I could see how this could loop hundreds of extra times.

Also, is there a case (specifically an exotic fee structure) where the initial estimate of `orderQuantity` will be less than the desired result? I'd like to see how this does with a progressive fee structure, maybe something like, less than 1k shares is $1 per share, but more than 1k shares is 1c per share. I know this is highly unlikely, but the iteration logic should be able to handle it. Before it was based on newton rhapson root finding, but this only decrements, may need to increment as well to be robust. Thoughts?